modules/rpc/yarpc.go

```diff
@@ -36,67 +35,192 @@ import (
 	tch "go.uber.org/yarpc/transport/tchannel"
 )
 
-// YarpcModule is an implementation of a core module using YARPC
-type YarpcModule struct {
+// YARPCModule is an implementation of a core RPC module using YARPC.
+// All the YARPC modules share the same dispatcher and middleware.
+// Dispatcher will start when any created module calls Start().
+type YARPCModule struct {
 	modules.ModuleBase
-	rpc                      *yarpc.Dispatcher
-	register                 registerServiceFunc
-	config                   yarpcConfig
-	log                      ulog.Log
-	stateMu                  sync.RWMutex
-	inboundMiddlewares       []middleware.UnaryInbound
-	onewayInboundMiddlewares []middleware.OnewayInbound
+	register  registerServiceFunc
+	config    yarpcConfig
+	log       ulog.Log
+	stateMu   sync.RWMutex
+	isRunning bool
 }
 
 var (
-	_dispatcherFn = defaultYarpcDispatcher
+	_dispatcherFn = defaultYARPCDispatcher
+	_dispatcherMu sync.Mutex
 
-	_ service.Module = &YarpcModule{}
+	_ service.Module = &YARPCModule{}
+
+	// A collection of all YARPC configs that are stored together to create
+	// a shared dispatcher. The YARPC team advised it to be a 'singleton'
+	// to control the lifecycle of all of the in/out bound traffic.
+	_configs configCollection
 )
 
-type registerServiceFunc func(module *YarpcModule)
+type registerServiceFunc func(module *YARPCModule)
 
 type yarpcConfig struct {
 	modules.ModuleConfig
-	Bind          string `yaml:"bind"`
-	AdvertiseName string `yaml:"advertiseName"`
+	Bind                    string
+	AdvertiseName           string `yaml:"advertiseName"`
+	inboundMiddleware       []middleware.UnaryInbound
+	onewayInboundMiddleware []middleware.OnewayInbound
+	inbounds                []transport.Inbound
+}
+
+// Stores a collection of all modules configs with a shared dispatcher
+// that are safe to call from multiple go routines. All the configs must
+// share the same AdvertiseName and represent a single service.
+type configCollection struct {
```

Comment:
can we call this something concrete? It is doing a lot more things than just collecting yarpc configurations. `rpcController` or `dispatcherManager`. configCollection seems pretty generic to have a dispatcher in it.